### DISCUSSION TITLE
Feature Suggestion: Bookmarks

### DISCUSSION THREAD

**Discussion by NorthhtroN** [advplyr/audiobookshelf#115] [Ideas]
Hi @advplyr, I want to say thank you first for creating this app, It has filled a much needed hole in my self hosting setup and has become one of my most used apps.

After using for a while I think a useful feature would be to allow users to create a "custom" bookmark for where they left off. The App's built in progress saving works well, but have had some instances where it may not have synced my last progress and I am having to jump forward/manually search through until I find where I left off.

**Comment by advplyr** [comment]
I think this is a great feature idea. I have also come across some sync issues here and there, particularly with the mobile app, that I hope to resolve.
I see bookmarks functioning similar to chapters. An icon that opens up a pop-up list of timestamps with an "add bookmark" button. Bookmarks could also be visible under the track like chapters but with a different color tick mark.

**Reply by Jdiesel87** [reply]
If you do end up adding this having the ability to name or add a note with the bookmark could be really nice to have. I see it as a way to document favorite parts of a book or areas that you might want to go back to a later time easily. Being able to to identify what they are would be extra helpful.

**Comment by advplyr** [comment]
I just pushed the very start of this feature in `v1.5.2` under experimental features.

There are some things incomplete so far:
- Need to add a way to remove bookmarks
- Bookmarks should be sorted ascending
- Bookmarks are currently stored with audiobook progress, which means it will be deleted if you remove progress. Need to repackage the progress data to serve more as a general user audiobook data.

I wasn't sure what this should look like, so I am hoping for some feedback on the design.

![image](https://user-images.githubusercontent.com/67830747/138617041-122e05a6-4d17-42a3-bd12-66837e0cc0b8.png)

![image](https://user-images.githubusercontent.com/67830747/138617156-86aa8e35-7a8b-4e04-9a29-0f72db0d1f58.png)

![image](https://user-images.githubusercontent.com/67830747/138617117-e8336779-a821-49a4-9217-1fecb05e4d2d.png)

**Reply by NorthhtroN** [reply]
That is a quick turn around! I'll update and give it some feedback through the week.

**Reply by NorthhtroN** [reply]
Took my first pass at running through this morning, from a design perspective I think you nailed it. easy to use, and functions how I would have thought. Matches up with how I have seen in other apps
My main feedback from a "design" side would be to add a default note for any bookmark, Maybe a date/time stamp. Other then that look a

After that we start to get more into functionality which I know is still a work in progress.
1) Allow edit to bookmark note
2) Potential user defined setting to start play from a bookmark x seconds before bookmark time (i,e book mark set for 1.21:19, start playing at 1:21:14) to let a listener "remember" where they left off before picking up right where they left off (May apply for resume play in any situation)

**Reply by advplyr** [reply]
Forgot about edit. I think your 2nd suggestion is a good first user-specific setting. Right now when you resume progress on an audiobook it picks up exactly where you left off, when it should probably jump back a little bit. I'm not sure if this should be adjustable in settings or if it should just be set to 5 seconds. If it is a setting then maybe the bookmark uses that settings.

**Reply by Jdiesel87** [reply]
Might be worth making it a universal behavior not exclusive to just bookmarks. I often resume a book and skip back a bit to reestablish my place in the book. Having the option to resume playback at -5, -10, -15, etc. intervals could be useful.

**Reply by advplyr** [reply]
I think it should only jump you back if you last paused longer than some amount of time.
for example I press pause then 3 seconds later hit play again I don't want to jump back 5 seconds even if I have settings set to that. 

A small but notable detail

**Reply by Jdiesel87** [reply]
Yeah good call. I was thinking more with stop/play than with pause/resume

**Reply by NorthhtroN** [reply]
I would agree, A global setting that applied to pause/bookmarks/other future features would be the suggestion, more then that would be over-complicating a design.

The detail on only jumping back after a set pause duration would be a nice polish on the setting as well.

**Comment by advplyr** [comment]
Another update to bookmarks in [v1.5.5](https://github.com/advplyr/audiobookshelf/releases/tag/v1.5.5)

Bookmarks can be edited and deleted. Also, create will not be shown if you already created a bookmark at the current time. I didn't think it would be useful to create multiple bookmarks at the same time, and it saves me from having to create an id for them. The bookmark id is just the time.

The last item remaining before this can be removed from experimental is to repurpose the AudiobookProgress data into a general UserAudiobook object.

The seek back feature discussed will be held off until this is implemented across the board. In PR #150 the start of seek back support was added to the android app for streams.

**Reply by advplyr** [reply]
Just pushed `v1.5.7` that finalized the last item. Bookmarks will be moved to production in the next release with a little more testing.

**Comment by advplyr** [comment]
Bookmarks are no longer experimental in release [v1.6.0](https://github.com/advplyr/audiobookshelf/releases/tag/v1.6.0)

All feedback & suggestions can still be put in this discussion.

**Comment by advplyr** [comment]
Bookmarks are now accessible on the mobile app in `v0.9.17-beta` in the new audio player.

**Comment by juriroemer** [comment]
Sorry I only got back at this just now - how would you feel about somewhat simplifying the UI? I think the extra view for the text field is somewhat unnecessary, it just adds another step. I also feel like the timestamp is the defining part and should be the focus on the left, instead of the bookmark icon (which gets redundant and in the end doesn't really serve any purpose).

Here's a mockup, what do you think?


![mockup](https://user-images.githubusercontent.com/31971585/140160248-800a68f6-a0eb-4357-84ca-4225ee0bf905.png)





https://user-images.githubusercontent.com/31971585/140162339-15540d9b-9afd-4a59-805d-d8f1e9bcd68d.mov

Timestamps are `text-gray-400`, the horizontal divider is `#555`, other than that it's pretty much exactly the same elements. Some kind of hover indicator (like a very slightly lighter blue) might be nice for the `+` button.

Also I think there's some kind of bug right now, where I can only create one bookmark at a time?


https://user-images.githubusercontent.com/31971585/140166180-bea3a5cf-9fd7-4deb-98b9-faa3a93f726e.mov

**Reply by advplyr** [reply]
Agree, that looks better. It isn't a bug it was intentional because you already have a bookmark for that exact time. That is why the icon is green, because you are already at that position in the audiobook.

It looks weird because the button is just not there at all, but wasn't sure how to handle that.

You also didn't show what edit looks like, I guess editing will just be editing inline and replacing the edit/delete icons with a submit button.

**Reply by juriroemer** [reply]
Oh that totally makes sense, didn't think of that. Maybe we can just have a disabled/gray text field and `+` button if there already is a bookmark at that position?

Yeah I agree inline would be a solid way to handle edits (as long as this doesn't push other elements around when the text is replaced with the (dimensionally larger) textfield). Maybe also have a `cancel` / `✘` button with the `submit` / `✔` button. And also map [Enter] and [esc] to `submit` and `cancel`.

**Reply by advplyr** [reply]
Just pushed `v1.6.3` with the updated bookmark design. I think I got most of it, except I still kept the create new hidden.
I think we will need to come up with a better way to show that you cannot create a bookmark when you already have a bookmark at that time.

![bookmarks_edit](https://user-images.githubusercontent.com/67830747/140229096-615cab29-0095-4ee6-9637-a82edbffe8d2.gif)

If you don't enter a note when creating new, then the default note will be a datetime string.

**Reply by juriroemer** [reply]
Cool!

Unfortunatly the title and text field seems fixed, which gives problems with longer timestamps :/ I guess a safe bet would be that the maximum timestamp length would be `xx:xx:xx` or `xxx:xx:xx` (so 8 or 9 characters), if the title/edit field has to have a fixed width?


![](https://user-images.githubusercontent.com/31971585/140231339-4250ee4c-06fb-4ca2-b672-f9ae401f30bd.png)
![](https://user-images.githubusercontent.com/31971585/140231342-13cb6ed9-8d71-444c-8d17-714f12b88806.png)

Edit: Also I think having the timestamp in a darker color like `text-gray-400` makes it easier to read and to distinguish between timestamp and note

**Reply by juriroemer** [reply]
best to align the notes of existing bookmarks with the "new bookmark" text field

![](https://user-images.githubusercontent.com/31971585/140234000-35edb5b8-9358-46ca-8d3b-89cf09e39307.png)

**Reply by advplyr** [reply]
Just released `v1.6.4` with a fix for that.

There is another issue where if you start an audiobook for the first time and you have no previous progress the bookmark will fail until some current time is saved. This is an easy fix I will do in the next update, but not that big of a deal.
Edit: Disregard, pushing a quick patch on this.
